python_modules/dagster/dagster/_core/definitions/asset_automation_evaluator.py

```diff
@@ -0,0 +1,258 @@
+
+    def __invert__(self) -> "AutomationCondition":
+        # convert a negated OrAutomationCondition into a NorAutomationCondition
+        if isinstance(self, OrAutomationCondition):
+            return NorAutomationCondition(children=self.children)
+        # convert a negated NorAutomationCondition into an OrAutomationCondition
+        elif isinstance(self, NorAutomationCondition):
+            return OrAutomationCondition(children=self.children)
+        return NorAutomationCondition(children=[self])
+
+
+class RuleCondition(
+    AutomationCondition, NamedTuple("_RuleCondition", [("rule", AutoMaterializeRule)])
+):
+    """This class represents the condition that a particular AutoMaterializeRule is satisfied."""
+
+    def evaluate(self, context: RuleEvaluationContext) -> ConditionEvaluation:
+        context.daemon_context._verbose_log_fn(f"Evaluating rule: {self.rule.to_snapshot()}")  # noqa
+        results = self.rule.evaluate_for_asset(context)
+        true_subset = context.empty_subset()
+        for _, asset_partitions in results:
+            true_subset |= AssetSubset.from_asset_partitions_set(
+                context.asset_key, context.partitions_def, asset_partitions
+            )
+        context.daemon_context._verbose_log_fn(f"Rule returned {true_subset.size} partitions")  # noqa
+        return ConditionEvaluation(condition=self, true_subset=true_subset, results=results)
+
+
+class AndAutomationCondition(
+    AutomationCondition,
+    NamedTuple("_AndAutomationCondition", [("children", Sequence[AutomationCondition])]),
+):
+    """This class represents the condition that all of its children evaluate to true."""
+
+    def evaluate(self, context: RuleEvaluationContext) -> ConditionEvaluation:
+        child_evaluations: List[ConditionEvaluation] = []
+        true_subset = context.candidate_subset
+        for child in self.children:
+            context = context.with_candidate_subset(true_subset)
+            result = child.evaluate(context)
+            child_evaluations.append(result)
+            true_subset &= result.true_subset
+        return ConditionEvaluation(
+            condition=self, true_subset=true_subset, children=child_evaluations
+        )
+
+
+class OrAutomationCondition(
+    AutomationCondition,
+    NamedTuple("_OrAutomationCondition", [("children", Sequence[AutomationCondition])]),
+):
+    """This class represents the condition that any of its children evaluate to true."""
+
+    def evaluate(self, context: RuleEvaluationContext) -> ConditionEvaluation:
+        child_evaluations: List[ConditionEvaluation] = []
+        true_subset = context.empty_subset()
+        for child in self.children:
+            result = child.evaluate(context)
+            child_evaluations.append(result)
+            true_subset |= result.true_subset
+        return ConditionEvaluation(
+            condition=self, true_subset=true_subset, children=child_evaluations
+        )
+
+
+class NorAutomationCondition(
```

Comment:
There will be situations where it's useful for users to expect the condition object tree.  E.g. you might want to print out a `__repr__` for debugging or traverse the tree to check whether programmatically. In those situations, I think it might be somewhat confusing / unexpected to see a "nor" show up when all you did was use `~`.

I.e. I would expect `x & ~y` to translate to `AndAutomationCondition([XAutomationCondition(), NotAutomationCondition(YAutomationCondition())])`, not `AndAutomationCondition([XAutomationCondition(), NorAutomationCondition([YAutomationCondition()])])`